File: simulator/src/java/darkness/simulator/dmx/BulbManager.java
package darkness.simulator.dmx;

import com.jme3.scene.Node;
import darkness.simulator.graphics.Point;

import java.util.Collection;
import java.util.HashMap;
import java.util.NoSuchElementException;

public class BulbManager {
    private static BulbManager instance = new BulbManager();

    private final HashMap<Integer, BulbRGB> bulbMap = new HashMap<Integer, BulbRGB>();
    private final HashMap<String, BulbRGB>  bulbNameMap = new HashMap<String, BulbRGB>();

    private BulbManager() {
    }

    public static BulbManager getInstance() { return instance; }

    public BulbRGB getBulb(int id) {
        return bulbMap.get(id);
    }

    public BulbRGB getBulb(String name) {
        return bulbNameMap.get(name);
    }

    /**
     * Creates a bulb with the given integer id and registers it
     */
    public BulbRGB registerBulb(int id, Channel channelRed, Channel channelGreen, Channel channelBlue,
            Point position, Node parentNode) {
        if(bulbMap.containsKey(id)) {
            throw new IllegalArgumentException("The bulb with the ID " + id + " already exists.");
        }

        BulbRGB bulb = new BulbRGB(channelRed, channelGreen, channelBlue, position, parentNode);
        bulbMap.put(id, bulb);
        return bulb;
    }


    public BulbRGB nameBulb(BulbRGB bulb, String name) {
        if(bulbNameMap.containsKey(name)) {
            throw new IllegalArgumentException("There is already a bulb with the name: "+name);
        }
        bulbNameMap.put(name, bulb);
        return  bulb;
    }

    public BulbRGB nameBulb(int id, String name) {
        if(!bulbMap.containsKey(id)) {
            throw new NoSuchElementException("There is no bulb with the ID: "+id);
        }
        return nameBulb(getBulb(id), name);
    }

    public Collection<BulbRGB> getAllBulbs() {
        return bulbMap.values();
    }
}

File: simulator/src/kotlin/darkness/simulator/dmx/BulbManager.kt
package darkness.simulator.dmx

import com.jme3.scene.Node
import darkness.simulator.graphics.Point
import java.util.NoSuchElementException

object BulbManager {
    private val bulbMap = mutableMapOf<Int, BulbRGB>()
    private val bulbNameMap = mutableMapOf<String, BulbRGB>()

    val allBulbs: Collection<BulbRGB>
        get() = bulbMap.values

    fun getBulb(id: Int): BulbRGB {
        return bulbMap[id]!!
    }

    fun getBulb(name: String): BulbRGB {
        return bulbNameMap[name]!!
    }

    /**
     * Creates a bulb with the given integer id and registers it
     */
    fun registerBulb(id: Int, channelRed: Channel, channelGreen: Channel, channelBlue: Channel,
                     position: Point, parentNode: Node): BulbRGB {
        if (bulbMap.containsKey(id)) {
            throw IllegalArgumentException("The bulb with the ID $id already exists.")
        }

        val bulb = BulbRGB(channelRed, channelGreen, channelBlue, position, parentNode)
        bulbMap[id] = bulb
        return bulb
    }


    fun nameBulb(bulb: BulbRGB, name: String): BulbRGB {
        if (bulbNameMap.containsKey(name)) {
            throw IllegalArgumentException("There is already a bulb with the name: $name")
        }
        bulbNameMap[name] = bulb
        return bulb
    }

    fun nameBulb(id: Int, name: String): BulbRGB {
        if (!bulbMap.containsKey(id)) {
            throw NoSuchElementException("There is no bulb with the ID: $id")
        }
        return nameBulb(getBulb(id), name)
    }
}
